### PR TITLE
[move-prover] Mutation testing framework

### PR DESCRIPTION
**Pull request by Checkmate50**
## Motivation

This PR will introduce a simple mutation testing framework to the move prover.  The idea behind this framework is to make testing a large number of mutations across a module feasible and to report surviving mutations.  The current status of the mutation framework is that it seems to work for single modules, but reports strange results for multiple modules.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Testing is done manually on example modules, notably [the arithm test](https://github.com/diem/diem/blob/main/language/move-prover/tests/sources/functional/arithm.move) (modified to not be failing before testing).

## Related PRs

Extends [this PR](https://github.com/diem/diem/pull/8725)

### CODE REVIEW

File: language/move-prover/mutation/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+#![forbid(unsafe_code)]
+
+use itertools::Itertools;
+use prover_mutation::mutator;
+
+fn main() {
+    let args = std::env::args().collect_vec();
+    if args.len() > 1 && matches!(args[1].as_str(), "-h" | "--help") {
+        println!("mutation: optionally select a subset of mutations to perform");
```
Comment (wrwg) [inline language/move-prover/mutation/src/main.rs:12]: I think --help is already implemented by the command line parser. It exits in this case, and prints options with all documentation.

File: language/move-prover/mutation/src/mutator.rs [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+        Ok(mutated)
+    }
+
+    fn mutate_module(&mut self, module: ModuleEnv<'_>) -> anyhow::Result<()> {
+        print!("running module {} ..", module.get_full_name_str());
+        std::io::stdout().flush()?;
+
+        // Scope verification to the given module
+        self.options.prover.verify_scope =
+            VerificationScope::OnlyModule(module.get_full_name_str());
```
Comment (wrwg) [inline language/move-prover/mutation/src/mutator.rs:185]: For followup PR: we want to mutate function by function, as discussed.

File: language/move-prover/mutation/src/mutator.rs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    Ok(())
+}
+
+impl Runner {
+    fn mutate(&mut self, env: &GlobalEnv) -> anyhow::Result<bool> {
+        let mut mutated = false;
+        for module in env.get_modules() {
+            if module.is_target() {
+                self.mutate_module(module)?;
+                let mm = env.get_extension::<MutationManager>();
```
Comment (wrwg) [inline language/move-prover/mutation/src/mutator.rs:166]: nit: `env.get_extension::<MutationManager>().map(|e| e.mutated).unwrap_or(false)`

File: language/move-prover/mutation/src/mutator.rs [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            &mut self.error_writer,
+            "unexpected boogie generation errors",
+        )?;
+
+        // Verify boogie, measuring duration.
+        let now = Instant::now();
+        verify_boogie(&env, &self.options, &targets, code_writer)?;
+
+        // Determine result status.
+        let status = if env.error_count() > 0 {
```
Comment (wrwg) [inline language/move-prover/mutation/src/mutator.rs:220]: For followup PR: this does not "kill" mutants yet.

**Review by wrwg — DISMISSED** [review]

**Comment by mengxu-fb** [comment]
I added a `bors-squash` label as we try to enforce the policy that every single commit should pass all tests on its own. `bors-squash` will combine all commits in this PR into one.

File: language/move-prover/mutation/src/main.rs (hunk 1 above)
Comment (Checkmate50) [reply language/move-prover/mutation/src/main.rs]: Fixed!

File: language/move-prover/mutation/src/mutator.rs (hunk 3 above)
Comment (Checkmate50) [reply language/move-prover/mutation/src/mutator.rs]: Updated, thanks for the tip!

**Comment by Checkmate50** [comment]
So I think this is ready to land, but I want to confirm there aren't any changes needed for this PR before moving onto the next phase

**Review by wrwg — APPROVED** [review]

**Pull request merged**